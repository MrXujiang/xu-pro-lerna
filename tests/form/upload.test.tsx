import React from 'react';
import ProForm, { ProFormUploadButton, ProFormUploadDragger } from '@ant-design/pro-form';
import { mount } from 'enzyme';
import mock from 'xhr-mock';
import { act } from 'react-dom/test-utils';
import { waitTime, waitForComponentToPaint } from '../util';

const mockFile = new File(['foo'], 'foo.png', {
  type: 'image/png',
});

export function setup() {
  mock.setup();
  // @ts-ignore
  mock.post('http://upload.com/', (req, res) => {
    req.headers({
      'content-length': '100',
    });
    req.body('thisisbody');
    return res;
  });
}

export const teardown = mock.teardown.bind(mock);

describe('ProFormUpload', () => {
  const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

  beforeEach(() => setup());
  afterEach(() => {
    teardown();
    errorSpy.mockReset();
  });

  it('🏐 ProFormUploadButton support onChange', async () => {
    const fn = jest.fn();
    const wrapper = mount(
      <ProForm
        onValuesChange={(_, values) => {
          fn(values.files);
        }}
      >
        <ProFormUploadButton
          action="http://upload.com"
          listType="text"
          label="upload"
          name="files"
        />
      </ProForm>,
    );

    act(() => {
      wrapper.find('.ant-upload input').simulate('change', {
        target: {
          files: [mockFile],
        },
      });
    });
    await waitTime(200);
    expect(fn).toBeCalled();
  });

  it('🏐 ProFormUploadButton support beforeUpload', async () => {
    const fn = jest.fn();
    const wrapper = mount(
      <ProForm
        onValuesChange={(_, values) => {
          fn(values.files);
        }}
      >
        <ProFormUploadButton
          action="http://upload.com"
          listType="text"
          label="upload"
          name="files"
          fieldProps={{
            beforeUpload: () => {
              return false;
            },
          }}
        />
      </ProForm>,
    );

    act(() => {
      wrapper.find('.ant-upload input').simulate('change', {
        target: {
          files: [mockFile],
        },
      });
    });
    await waitTime(200);

    act(() => {
      expect(wrapper.find('div.ant-upload-list-picture-container').length).toBe(0);
    });
  });

  it('🏐 ProFormUploadButton support disable', async () => {
    const wrapper = mount(
      <ProFormUploadButton
        disabled
        action="http://upload.com"
        listType="text"
        label="upload"
        name="files"
      />,
    );
    expect(wrapper.find('Upload Button')).toMatchSnapshot();
    act(() => {
      wrapper.setProps({
        buttonProps: {
          disabled: true,
          type: 'dashed',
        },
      });
    });
    await waitForComponentToPaint(wrapper);
    expect(wrapper.find('Upload Button')).toMatchSnapshot();
    act(() => {
      wrapper.setProps({
        disabled: false,
        buttonProps: {},
        fieldProps: {
          disabled: true,
        },
      });
    });
    await waitForComponentToPaint(wrapper);
    expect(wrapper.find('Upload Button')).toMatchSnapshot();
  });

  it('🏐 ProFormUploadDragger support onChange', async () => {
    const fn = jest.fn();
    const onChangeFn = jest.fn();
    const wrapper = mount(
      <ProForm
        onValuesChange={(_, values) => {
          fn(values.files);
        }}
      >
        <ProFormUploadDragger
          onChange={() => onChangeFn()}
          action="http://upload.com"
          label="upload"
          name="files"
        />
      </ProForm>,
    );

    act(() => {
      wrapper.find('.ant-upload input').simulate('change', {
        target: {
          files: [mockFile],
        },
      });
    });
    await waitTime(200);
    expect(fn).toBeCalled();
    expect(onChangeFn).toBeCalled();
  });

  it('🏐 ProFormUploadDragger hide when max', async () => {
    const wrapper = mount(
      // @ts-ignore
      <ProFormUploadDragger
        max={2}
        value={[mockFile, mockFile, mockFile]}
        action="http://upload.com"
        label="upload"
        name="files"
      />,
    );

    await waitTime(200);
    expect(wrapper.find('.ant-upload.ant-upload-drag').props().style?.display).toBe('none');
  });

  it('🏐 ProFormUploadDragger support children', async () => {
    const extra = 'extra';
    const wrapper = mount(
      // @ts-ignore
      <ProFormUploadDragger
        value={[mockFile, mockFile, mockFile]}
        action="http://upload.com"
        label="upload"
        name="files"
      >
        {extra}
      </ProFormUploadDragger>,
    );

    await waitTime(200);
    expect(wrapper.find('.ant-upload-drag .ant-upload-extra').first().text()).toBe(extra);
  });

  it('🏐 ProFormUploadButton hide when max', async () => {
    const wrapper = mount(
      // @ts-ignore
      <ProFormUploadButton
        max={2}
        value={[mockFile, mockFile, mockFile]}
        action="http://upload.com"
        label="upload"
        name="files"
      />,
    );

    await waitTime(200);
    expect(wrapper.find('.anticon.anticon-upload').exists()).toBe(false);
  });
});
