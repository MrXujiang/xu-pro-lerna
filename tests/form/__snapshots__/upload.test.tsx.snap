// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProFormUpload üèê ProFormUploadButton support disable 1`] = `
<Button
  disabled={true}
>
  <Wave>
    <button
      className="ant-btn"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <LoadingIcon
        existIcon={false}
        loading={false}
        prefixCls="ant-btn"
      >
        <CSSMotion
          motionName="ant-btn-loading-icon-motion"
          onAppearActive={[Function]}
          onAppearStart={[Function]}
          onEnterActive={[Function]}
          onEnterStart={[Function]}
          onLeaveActive={[Function]}
          onLeaveStart={[Function]}
          removeOnLeave={true}
          visible={false}
        >
          <DomWrapper />
        </CSSMotion>
      </LoadingIcon>
      <ForwardRef(UploadOutlined)
        key=".0"
      >
        <AntdIcon
          icon={
            Object {
              "icon": Object {
                "attrs": Object {
                  "focusable": "false",
                  "viewBox": "64 64 896 896",
                },
                "children": Array [
                  Object {
                    "attrs": Object {
                      "d": "M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z",
                    },
                    "tag": "path",
                  },
                ],
                "tag": "svg",
              },
              "name": "upload",
              "theme": "outlined",
            }
          }
        >
          <span
            aria-label="upload"
            className="anticon anticon-upload"
            role="img"
          >
            <IconReact
              icon={
                Object {
                  "icon": Object {
                    "attrs": Object {
                      "focusable": "false",
                      "viewBox": "64 64 896 896",
                    },
                    "children": Array [
                      Object {
                        "attrs": Object {
                          "d": "M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z",
                        },
                        "tag": "path",
                      },
                    ],
                    "tag": "svg",
                  },
                  "name": "upload",
                  "theme": "outlined",
                }
              }
            >
              <svg
                aria-hidden="true"
                data-icon="upload"
                fill="currentColor"
                focusable="false"
                height="1em"
                key="svg-upload"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                  key="svg-upload-svg-0"
                />
              </svg>
            </IconReact>
          </span>
        </AntdIcon>
      </ForwardRef(UploadOutlined)>
      <span
        key=".1"
      >
        ÂçïÂáª‰∏ä‰º†
      </span>
    </button>
  </Wave>
</Button>
`;

exports[`ProFormUpload üèê ProFormUploadButton support disable 2`] = `
<Button
  disabled={true}
  type="dashed"
>
  <Wave>
    <button
      className="ant-btn ant-btn-dashed"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <LoadingIcon
        existIcon={false}
        loading={false}
        prefixCls="ant-btn"
      >
        <CSSMotion
          motionName="ant-btn-loading-icon-motion"
          onAppearActive={[Function]}
          onAppearStart={[Function]}
          onEnterActive={[Function]}
          onEnterStart={[Function]}
          onLeaveActive={[Function]}
          onLeaveStart={[Function]}
          removeOnLeave={true}
          visible={false}
        >
          <DomWrapper />
        </CSSMotion>
      </LoadingIcon>
      <ForwardRef(UploadOutlined)
        key=".0"
      >
        <AntdIcon
          icon={
            Object {
              "icon": Object {
                "attrs": Object {
                  "focusable": "false",
                  "viewBox": "64 64 896 896",
                },
                "children": Array [
                  Object {
                    "attrs": Object {
                      "d": "M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z",
                    },
                    "tag": "path",
                  },
                ],
                "tag": "svg",
              },
              "name": "upload",
              "theme": "outlined",
            }
          }
        >
          <span
            aria-label="upload"
            className="anticon anticon-upload"
            role="img"
          >
            <IconReact
              icon={
                Object {
                  "icon": Object {
                    "attrs": Object {
                      "focusable": "false",
                      "viewBox": "64 64 896 896",
                    },
                    "children": Array [
                      Object {
                        "attrs": Object {
                          "d": "M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z",
                        },
                        "tag": "path",
                      },
                    ],
                    "tag": "svg",
                  },
                  "name": "upload",
                  "theme": "outlined",
                }
              }
            >
              <svg
                aria-hidden="true"
                data-icon="upload"
                fill="currentColor"
                focusable="false"
                height="1em"
                key="svg-upload"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                  key="svg-upload-svg-0"
                />
              </svg>
            </IconReact>
          </span>
        </AntdIcon>
      </ForwardRef(UploadOutlined)>
      <span
        key=".1"
      >
        ÂçïÂáª‰∏ä‰º†
      </span>
    </button>
  </Wave>
</Button>
`;

exports[`ProFormUpload üèê ProFormUploadButton support disable 3`] = `
<Button
  disabled={true}
>
  <Wave>
    <button
      className="ant-btn"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <LoadingIcon
        existIcon={false}
        loading={false}
        prefixCls="ant-btn"
      >
        <CSSMotion
          motionName="ant-btn-loading-icon-motion"
          onAppearActive={[Function]}
          onAppearStart={[Function]}
          onEnterActive={[Function]}
          onEnterStart={[Function]}
          onLeaveActive={[Function]}
          onLeaveStart={[Function]}
          removeOnLeave={true}
          visible={false}
        >
          <DomWrapper />
        </CSSMotion>
      </LoadingIcon>
      <ForwardRef(UploadOutlined)
        key=".0"
      >
        <AntdIcon
          icon={
            Object {
              "icon": Object {
                "attrs": Object {
                  "focusable": "false",
                  "viewBox": "64 64 896 896",
                },
                "children": Array [
                  Object {
                    "attrs": Object {
                      "d": "M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z",
                    },
                    "tag": "path",
                  },
                ],
                "tag": "svg",
              },
              "name": "upload",
              "theme": "outlined",
            }
          }
        >
          <span
            aria-label="upload"
            className="anticon anticon-upload"
            role="img"
          >
            <IconReact
              icon={
                Object {
                  "icon": Object {
                    "attrs": Object {
                      "focusable": "false",
                      "viewBox": "64 64 896 896",
                    },
                    "children": Array [
                      Object {
                        "attrs": Object {
                          "d": "M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z",
                        },
                        "tag": "path",
                      },
                    ],
                    "tag": "svg",
                  },
                  "name": "upload",
                  "theme": "outlined",
                }
              }
            >
              <svg
                aria-hidden="true"
                data-icon="upload"
                fill="currentColor"
                focusable="false"
                height="1em"
                key="svg-upload"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                  key="svg-upload-svg-0"
                />
              </svg>
            </IconReact>
          </span>
        </AntdIcon>
      </ForwardRef(UploadOutlined)>
      <span
        key=".1"
      >
        ÂçïÂáª‰∏ä‰º†
      </span>
    </button>
  </Wave>
</Button>
`;
